internal/supervisor/import.go

```diff
@@ -0,0 +1,394 @@
+	}
+
+	return nil
+}
+
+func (is *ImportSupervisor) getBucketForInstallation(installation *model.Installation) (string, error) {
+	switch installation.Filestore {
+	// TODO handle single tenant bucket names
+	case model.InstallationFilestoreMultiTenantAwsS3:
+		fallthrough
+	case model.InstallationFilestoreBifrost:
+		return toolsAWS.GetMultitenantBucketNameForInstallation(installation.ID, is.store, is.awsClient)
+	case model.InstallationFilestoreAwsS3:
+		fallthrough
+	case model.InstallationFilestoreMinioOperator:
+		return "", errors.Errorf("support for workspace imports to workspaces with the filestore type %s not yet supported", installation.Filestore)
+	default:
+		return "", errors.Errorf("encountered unknown filestore type %s", installation.Filestore)
+	}
+}
+
+// copyImportToWorkspaceFilestorecopies the archive from the AWAT's bucket to the Installation's
+// bucket. Returns the key of the archive in the source bucket or an
+// error
+func (is *ImportSupervisor) copyImportToWorkspaceFilestore(imprt *awat.ImportStatus, installation *model.Installation) (string, error) {
+	// calculate necessary paths and copy the archive to the Installation's S3 bucket
+	source := strings.SplitN(imprt.Resource, "/", 2)
+	if len(source) != 2 {
+		return "", errors.Errorf("failed to parse bucket/key from Import %s Resource %s", imprt.ID, imprt.Resource)
+	}
+	srcBucket := source[0]
+	srcKey := source[1]
+	destKey := fmt.Sprintf("%s/import/%s", installation.ID, srcKey)
+
+	destBucket, err := is.getBucketForInstallation(installation)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to determine bucket name for Installation %s", installation.ID)
+	}
+	is.logger.Debugf("copying %s/%s to %s/%s", srcBucket, srcKey, destBucket, destKey)
+	err = is.awsClient.S3LargeCopy(&srcBucket, &srcKey, &destBucket, &destKey)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to copy archive to Installation %s", installation.ID)
+	}
+
+	return srcKey, nil
+}
+
+func (is *ImportSupervisor) startImportProcessAndWait(mmctl *mmctl, importArchiveFilename string, awatImportID string) error {
+	output, err := mmctl.Do("import", "process", importArchiveFilename)
+	if err != nil {
+		return errors.Wrap(err, "failed to start import process in Mattermost itself")
+	}
+
+	jobResponses := []*jobResponse{}
+	err = json.Unmarshal([]byte(output), &jobResponses)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal Job response from Mattermost")
+	}
+
+	if len(jobResponses) < 1 {
+		return errors.New("response was empty")
+	}
+
+	jobID := jobResponses[0].ID
+	is.logger.Infof("Started Import job %s", jobID)
+	return is.waitForImportToComplete(mmctl, jobID, awatImportID)
+}
+
+func (is *ImportSupervisor) waitForImportToComplete(mmctl *mmctl, mattermostJobID string, awatImportID string) error {
+	complete := false
+	var (
+		resp         *jobResponse
+		output       []byte
+		jobResponses []*jobResponse
+	)
+	var err error = nil
+	defer func(is *ImportSupervisor) {
```

Comment:
Yes, it needs to be deferred or it needs to be called whenever this method is complete, because the AWAT should be notified when it is complete, regardless as to the result. To eliminate this `defer` we'd have to define a helper function and call it everywhere we `return` in this function manually, and then remember to call it in all new `return`s added to the function in the future. To me, that seems like the right time to use `defer`.

Reply:
Yeah I get that. Sorry my comment might have not been clear enough, what I am suggesting is to move deferred logic to the caller as there is only one place where the `waitForImportToComplete` is called. 

So I wonder if we could move the logic to `startImportProcessAndWait` without worrying about not non-fatal errors:
```go
func (is *ImportSupervisor) startImportProcessAndWait(mmctl *mmctl, importArchiveFilename string, awatImportID string) error {
    ...

    completeReq := &awat.ImportCompletedWorkRequest{
        ID:         awatImportID,
        CompleteAt: time.Now().UnixNano() / int64(time.Millisecond),
    })

    err := is.waitForImportToComplete(...)
    if err != nil {
    // log error
        completeReq.Error = err.Error()
    }
    err = is.awatClient.CompleteImport(completeReq)
    if err != nil {
        // log error
    }
    return nil
}
```

Reply:
Oh, I definitely misunderstood! Let me see if I can incorporate this suggestion; I get the feeling that it's a good one

Reply:
See commit 415516a for the resolution